File: src/LC_92.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} left
 * @param {number} right
 * @return {ListNode}
 */
// Iterative
var reverseBetween = function(head, left, right) {
  if (!head || left === right) return head;
  let dummy = new ListNode(0);
  dummy.next = head;
  let prev = dummy;

  for (let i = 0; i < left - 1; i++) {
    prev = prev.next;
  }
  let current = prev.next;
  let next = null;
  for (i = left; i < right; i++) {
    next = current.next;
    current.next = next.next;
    next.next = prev.next;
    prev.next = next;
  }
  return prev;
};

File: src/LC_92.java
class Solution {
  public ListNode reverseBetween(ListNode head, int left, int right) {
    if (head == null || left == right) return head;
    ListNode dummy = new ListNode(0);
    ListNode prev = dummy;
    dummy.next = head;
    for (int i = 0; i < left - 1; i++) {
        prev = prev.next;
    }
    ListNode current = prev.next;
    ListNode next = null;
    for (int i = left; i < right; i++) {
        next = current.next;
        current.next = next.next;
        next.next = prev.next;
        prev.next = next;
    }
    return dummy.next;
  }
}